### DISCUSSION TITLE
using brew test-bot with github actions

### DISCUSSION THREAD

**Discussion by ipatch** [Homebrew/discussions#1970] [Tap maintenance and Homebrew development]
i'm attempting to auto bottle formulae files contained with the homebrew-freecad repo, but before i go blowing up 💣 💥 that repo i've created a little testing ground (if you will) [homebrew-freecad-pg13](https://github.com/ipatch/homebrew-freecad-pg13). I've checked in one formula file ie. swig. i'm using the default workflow files when setting up a new homebrew tap. and they **don't** require much editing out of the box. 👍, however trying to build bottles for the last 3 major versions of macos ie.

- 10.14
- 10.15
- 11.0 (big sur, some may refer to it as 10.16)

i'm unable to build bottles for all three major OSes using _brew test-bot_ if i leave the workflow file as is, ie. `macos-latest` it appears there is no difference between `macos-latest` or `macOS-latest` my brew test-bot workflow completes then i can run the brew pr workflow to publish the bottle. (obviously triggering that workflow **after** the first workflow has completed).

however, using `macos-latest` only generates a bottle for **catalina** i would like to have bottles for all 3 above mentioned OSes.

is this presently possible with the current _brew test-bot_ setup. i know _homebrew-core_ provides bottles for all 3 OSes. so i'm just not sure what i'm missing here.

and i went poking around the homebrew-core repo, and saw mention of several runners, experimented with the ones defined below, but my jobs would never start.

https://github.com/Homebrew/homebrew-core/blob/master/.github/workflows/tests.yml#L77

---

basically, i'd to "auto" build bottles of the tap within the homebrew-freecad repo, and have bottles for mojave, catalina, and big sur if possible.

probably the only "documentation" i've really come across that explains the use of building bottle for a tap using github actions would be, https://brew.sh/2020/11/18/homebrew-tap-with-bottles-uploaded-to-github-releases/

and i didn't see much info / documentation regarding _test bot_ within the test bot repo itself.

**Comment by carlocab** [comment]
If you need to build bottles for Mojave and Big Sur, you'll probably need to provide your own self-hosted runners. Bottles built in Homebrew/core use self-hosted CI.

Using the default workflow will use GitHub runners, and GitHub doesn't provide Mojave runners, while their Big Sur runners are in private preview.

**Reply by ipatch** [reply]
i've tried using _self-hosted_ runner's in the past and did not have much luck with interfacing them with homebrew's github actions. i guess i'll keep doing my do diligence.  You wouldn't know of any resources, ie. blog post or what not how to use _brew test-bot_ with _self-hosted_ runners to setup auto bottling would you?

**Reply by carlocab** [reply]
Oops. I missed this. I assume you've worked it out by now. Because I wouldn't have been much help had I seen this much earlier. 😅

All I know is that you need to get your self-hosted runners wired up to GitHub, and the workflows should do the rest...

Maybe you can write that blogpost you were looking for?

**Reply by ipatch** [reply]
ehhh i'd be more interested in getting mojave support back for _self-hosted_ runners using brew test-bot. i added a comment on another post somewhere in this discussions, but don't remember exactly thread i commented on. somebody that's part of the homebrew-core org mentioned they'd take a look at it, but that was several months ago.

**Reply by carlocab** [reply]
@gromgit has been building Mojave bottles. Not sure how they've set it up though.

But I also don't see why having a self-hosted Mojave runner shouldn't work.

**Reply by ipatch** [reply]
> @gromgit has been building Mojave bottles. Not sure how they've set it up though.
> 
> But I also don't see why having a self-hosted Mojave runner shouldn't work.

yeah that's the guy, when attempting to build a bottle for mojave using test-bot / (the brew task) gets to the point where it detects that it is an unsupported version of macos, and probably because the `brew` command is doing a `brew update` or something of that nature where it's running a specific version of `brew` that no longer supports mojave and then the CI task stops.

**Reply by carlocab** [reply]
Probably `brew doctor`. If you get rid of that step from the workflow I think you should get on just fine.

**Reply by ipatch** [reply]
cool, that'd be really awesome. thanks for sharing that. i take a look at that in a few and see if it works.

**Reply by ipatch** [reply]
> Probably `brew doctor`. If you get rid of that step from the workflow I think you should get on just fine.

so just did a search through, https://github.com/FreeCAD/homebrew-freecad/blob/master/.github/workflows/tests.yml

and did not see a `brew doctor` cmd in there.

so i'd assume i'd have to fork _test-bot_ or something of that nature to disable `brew doctor`, or am i missing something else here?

**Reply by carlocab** [reply]
It is run implicitly by `--only-setup`. It should probably be safe to not do that step. I'll look into disabling the unsupported macOS error if you have `CI` set. Or skipping `brew doctor` if you're on an unsupported macOS -- one of the two.

**Reply by carlocab** [reply]
Yea, I suggest making that step
```
brew test-bot --only-setup || true
```
for now. You can always have a look at the log to double-check the `brew doctor` output, but in general I wouldn't worry too much about errors from there.

**Reply by ipatch** [reply]
i'll give it a go hear in a bit. thanks so much for following up. 🙏

**Reply by carlocab** [reply]
Actually, I just realised you could also just set `HOMEBREW_DEVELOPER=1` in your environment to avoid `brew doctor` failures due to the macOS version, in case you want the other warnings from `brew doctor`. See https://github.com/Homebrew/brew/blob/b28dba9861f4c2f41de1b013db313be57887ef0f/Library/Homebrew/extend/os/mac/diagnostic.rb#L108-L109

**Reply by gromgit** [reply]
> @gromgit has been building Mojave bottles. Not sure how they've set it up though.

I don't (currently) use GitHub Actions, mostly because I'm still trying to refine my seriously hacky core-merge workflow.

Also, given the daily tsunami of core tap updates, it just makes more sense to batch-update my Mojave tap, and batch-run speculative bottle builds locally, a certain number of which will fail.

**Reply by ipatch** [reply]
> Actually, I just realised you could also just set `HOMEBREW_DEVELOPER=1` in your environment to avoid `brew doctor` failures due to the macOS version, in case you want the other warnings from `brew doctor`. See [Homebrew/brew@`b28dba9`/Library/Homebrew/extend/os/mac/diagnostic.rb#L108-L109](https://github.com/Homebrew/brew/blob/b28dba9861f4c2f41de1b013db313be57887ef0f/Library/Homebrew/extend/os/mac/diagnostic.rb#L108-L109)

so i added a couple of lines to the `tests.yml` for the homebrew-freecad tap

https://github.com/FreeCAD/homebrew-freecad/blob/6fc0a11db4ed94edfa841cd2a5af749e1ee3c8c8/.github/workflows/tests.yml#L71

do you think this will allow the homebrew-freecad tap to build bottles for macos mojave?

**Reply by carlocab** [reply]
Ah, no, that's not how environment variables work in GitHub workflows. You can set a property called `env` earlier in the workflow to set environment variables, or do something like
```
echo HOMEBREW_DEVELOPER=1 >> $GITHUB_ENV
```
(Or something like that) Succeeding steps won't see this if all you do is `export`.

I'd probably just also set `HOMEBREW_DEVELOPER` unconditionally.

**Reply by ipatch** [reply]
thanks it been churning along for ~ 7 hours now, so when/if this job fails i'll update the `tests.yml` file.

**Comment by dpo** [comment]
If you replace `macos-latest` with ` macos-10.15, macos-11` in the list of OSes, you'll at least get Catalina and Big Sur bottles.
